diff --git a/connection.h b/connection.h
index 74dd413..f6619a0 100644
--- a/connection.h
+++ b/connection.h
@@ -48,6 +48,7 @@ struct ioabs;
 
 typedef struct _connection {
     int s;                  /* connected socket                 */
+    int s_index;            /* pfds index */
     struct sockaddr_in sin; /* name of peer                     */
     char *remote_ip;        /* ASCII remote IP address          */
     struct sockaddr_in sin_local; /* name of local side         */
diff --git a/ioabs_tcp.c b/ioabs_tcp.c
index 36d87d0..69fc5ff 100644
--- a/ioabs_tcp.c
+++ b/ioabs_tcp.c
@@ -78,13 +78,13 @@ static void ioabs_tcp_pre_select(connection c, int *n, struct pollfd *pfds) {
     struct ioabs_tcp *io;
     io = (struct ioabs_tcp*)c->io;
 
-    pfds[c->s].fd = c->s;
-    pfds[c->s].events |= POLLIN;
+    (*n)++;
+    c->s_index = *n;
+
+    pfds[c->s_index].fd = c->s;
+    pfds[c->s_index].events |= POLLIN;
     if (buffer_available(c->wrb) > 0)
-       pfds[c->s].events |= POLLOUT;
-    
-    if (c->s > *n)
-        *n = c->s;
+       pfds[c->s_index].events |= POLLOUT;
 }
 
 /* ioabs_tcp_post_select:
@@ -95,7 +95,7 @@ static int ioabs_tcp_post_select(connection c, struct pollfd *pfds) {
     struct ioabs_tcp *io;
     io = (struct ioabs_tcp*)c->io;
 
-    if (pfds[c->s].revents & (POLLIN | POLLHUP)) {
+    if (pfds[c->s_index].revents & (POLLIN | POLLHUP)) {
         /* Can read data. */
         do {
             char *r;
@@ -124,7 +124,7 @@ static int ioabs_tcp_post_select(connection c, struct pollfd *pfds) {
         }
     }
 
-    if (pfds[c->s].revents & POLLOUT && buffer_available(c->wrb) > 0) {
+    if (pfds[c->s_index].revents & POLLOUT && buffer_available(c->wrb) > 0) {
         /* Can write data. */
         n = 1;
         do {
diff --git a/ioabs_tls.c b/ioabs_tls.c
index 98ac7b7..dc47a7d 100644
--- a/ioabs_tls.c
+++ b/ioabs_tls.c
@@ -245,15 +245,15 @@ static void ioabs_tls_pre_select(connection c, int *n, struct pollfd *pfds) {
     struct ioabs_tls *io;
     io = (struct ioabs_tls*)c->io;
 
-    pfds[c->s].fd = c->s;
-    pfds[c->s].events |= POLLIN; /* always want to read */
+    (*n)++;
+    c->s_index = *n;
+
+    pfds[c->s_index].fd = c->s;
+    pfds[c->s_index].events |= POLLIN; /* always want to read */
     if (!io->write_blocked_on_read &&
         (buffer_available(c->wrb) > 0 || io->accept_blocked_on_write
          || io->read_blocked_on_write || io->shutdown_blocked_on_write))
-        pfds[c->s].events |= POLLOUT;
-
-    if (c->s > *n)
-        *n = c->s;
+        pfds[c->s_index].events |= POLLOUT;
 }
 
 /* ioabs_tls_post_select:
@@ -265,8 +265,8 @@ static int ioabs_tls_post_select(connection c, struct pollfd *pfds) {
     struct ioabs_tls *io;
     io = (struct ioabs_tls*)c->io;
 
-    canread  = pfds[c->s].revents & (POLLIN | POLLHUP);
-    canwrite = pfds[c->s].revents & POLLOUT;
+    canread  = pfds[c->s_index].revents & (POLLIN | POLLHUP);
+    canwrite = pfds[c->s_index].revents & POLLOUT;
     
     /* First, accept handling. */
     if ((io->accept_blocked_on_read && canread) || (io->accept_blocked_on_write && canwrite)) {
diff --git a/listener.h b/listener.h
index e4d5ad8..fefc300 100644
--- a/listener.h
+++ b/listener.h
@@ -51,6 +51,7 @@ typedef struct _listener {
     } tls;
 #endif
     int s;
+    int s_index;
 } *listener;
 
 /* the arguments of the constructor vary according to the particular
diff --git a/netloop.c b/netloop.c
index f39018b..e02f3e8 100644
--- a/netloop.c
+++ b/netloop.c
@@ -121,23 +121,26 @@ static void remove_connection(connection c) {
 /* listeners_pre_select:
  * Called before the main select(2) so listening sockets can be polled. */
 static void listeners_pre_select(int *n, struct pollfd *pfds) {
+    int i = 0;
     item *t;
     vector_iterate(listeners, t) {
         int s = ((listener)t->v)->s;
-       pfds[s].fd = s;
-       pfds[s].events |= POLLIN;
-        if (s > *n) *n = s;
+	pfds[i].fd = s;
+	pfds[i].events |= POLLIN;
+	((listener)t->v)->s_index = i;
+	if (i > *n) *n = i;
+	i++;
     }
 }
 
 /* listeners_post_select:
  * Called after the main select(2) to allow listening sockets to sort
  * themselves out. */
-static void listeners_post_select(struct pollfd *pfds) {
+static void listeners_post_select(struct pollfd *pfds, int *n) {
     item *t;
     vector_iterate(listeners, t) {
         listener L = (listener)t->v;
-       if (pfds[L->s].revents & (POLLIN | POLLHUP)) {
+       if (pfds[L->s_index].revents & (POLLIN | POLLHUP)) {
             struct sockaddr_in sin, sinlocal;
             size_t l = sizeof(sin);
             static int tcp_send_buf = -1;
@@ -549,6 +552,8 @@ void net_loop(void) {
     extern int child_died_signal;
     sigset_t chmask;
     struct pollfd *pfds;
+    int max_listeners;
+    item *t;
     
     sigemptyset(&chmask);
     sigaddset(&chmask, SIGCHLD);
@@ -557,7 +562,12 @@ void net_loop(void) {
     max_connections = 2 * max_running_children;
     connections = (connection*)xcalloc(max_connections, sizeof(connection*));
 
-    pfds = xmalloc(max_connections * sizeof *pfds);
+    /* find out number of listeners */
+    max_listeners = 0;
+    vector_iterate(listeners, t)
+	    max_listeners++;
+
+    pfds = xmalloc((max_listeners + max_connections) * sizeof *pfds);
 
     log_print(LOG_INFO, _("net_loop: tpop3d version %s successfully started"), TPOP3D_VERSION);
     
@@ -565,7 +575,7 @@ void net_loop(void) {
     while (!foad) {
         int n = 0, e, i;
 
-        for (i = 0; i < max_connections; ++i) {
+        for (i = 0; i < (max_listeners + max_connections); ++i) {
             pfds[i].fd = -1;
             pfds[i].events = pfds[i].revents = 0;
         }
@@ -579,7 +589,7 @@ void net_loop(void) {
             log_print(LOG_WARNING, "net_loop: poll: %m");
         } else if (e >= 0) {
             /* Check for new incoming connections */
-            if (!post_fork) listeners_post_select(pfds);
+            if (!post_fork) listeners_post_select(pfds, &n);
 
             /* Monitor existing connections */
             connections_post_select(pfds);
