commit 3e9deb3864be47369f6bcb86ca2733cbd208212d
Author: Arkadiusz Miśkiewicz <arekm@maven.pl>
Date:   Sun Jun 6 12:35:31 2010 +0200

    Look for libcrypto, too (needed for SHA1* functions).

diff --git a/configure.in b/configure.in
index db64849..1c0c2de 100644
--- a/configure.in
+++ b/configure.in
@@ -588,6 +588,7 @@ fi
 if test x"$enable_tls" = x"yes" || test x"$enable_sha1_passwords" = x"yes"
 then
     AC_CHECK_LIB2(ssl, SSL_new, [], AC_MSG_ERROR([[TLS/SHA1 passwords enabled, but can't find libssl]]), -lcrypto)
+    AC_CHECK_LIB(crypto, SHA1_Init, [], AC_MSG_ERROR([[TLS/SHA1 passwords enabled, but can't find libcrypto]]) )
 fi
 
 # GNU dbm
diff -urN tpop3d-1.5.5.org/configure.in tpop3d-1.5.5/configure.in
--- tpop3d-1.5.5.org/configure.in	2008-07-12 17:12:39.000000000 +0200
+++ tpop3d-1.5.5/configure.in	2013-06-09 18:55:01.370497207 +0200
@@ -11,7 +11,7 @@
 
 AC_INIT(auth_mysql.c)
 AC_CONFIG_AUX_DIR(config)
-AM_CONFIG_HEADER(configuration.h)
+AC_CONFIG_HEADERS(configuration.h)
 AM_INIT_AUTOMAKE(tpop3d, "1.5.5")
 
 AC_DEFINE_UNQUOTED(TPOP3D_VERSION, "$VERSION", [The tpop3d version number])
