diff -urN tpop3d-1.5.3.org/authswitch.c tpop3d-1.5.3/authswitch.c
--- tpop3d-1.5.3.org/authswitch.c	2004-01-28 20:59:07.879436352 +0100
+++ tpop3d-1.5.3/authswitch.c	2004-01-28 20:59:38.938970374 +0100
@@ -330,7 +330,7 @@
 #ifdef USE_WHOSON
     char buf[128] = {0};
     /* Notify whoson server the user has logged in correctly */
-    if (wso_login(clienthost, A->user, buf, sizeof(buf)) == -1)
+    if (whoson_enable && wso_login(clienthost, A->user, buf, sizeof(buf)) == -1)
         log_print(LOG_ERR, "authswitch_onlogin: wso_login: %s", buf);
 #endif /* USE_WHOSON */
     
