diff -urN tpop3d-1.5.3.org/netloop.c tpop3d-1.5.3/netloop.c
--- tpop3d-1.5.3.org/netloop.c	2006-01-24 17:22:20.730963000 +0100
+++ tpop3d-1.5.3/netloop.c	2006-01-24 17:23:01.710963000 +0100
@@ -495,9 +495,9 @@
                     if ((p = connection_parsecommand(c)) && p->cmd == QUIT)
                         c->m->apply_changes(c->m);
                 }
-                log_print(LOG_INFO, _("connections_post_select: client %s: finished session for `%s' with %s"), c->idstr, c->a->user, c->a->auth);
+                log_print(LOG_NOTICE, _("connections_post_select: client %s: finished session for `%s' with %s"), c->idstr, c->a->user, c->a->auth);
             }
-            log_print(LOG_INFO, _("connections_post_select: client %s: disconnected; %d/%d bytes read/written"), c->idstr, c->nrd, c->nwr);
+            log_print(LOG_NOTICE, _("connections_post_select: client %s: disconnected; %d/%d bytes read/written"), c->idstr, c->nrd, c->nwr);
 
 /*            remove_connection(c);*/
             connections[i] = NULL;
