? tpop3d-mysql-reconnect.patch
Index: auth_mysql.c
===================================================================
RCS file: /home/chris/vcvs/repos/tpop3d/auth_mysql.c,v
retrieving revision 1.56
diff -u -u -r1.56 auth_mysql.c
--- auth_mysql.c	19 Aug 2004 23:57:05 -0000	1.56
+++ auth_mysql.c	14 Dec 2006 14:47:11 -0000
@@ -85,6 +85,7 @@
     static MYSQL mysql_handle;
     char *password;
     unsigned int timeout;
+    my_bool my_true = 1;
 
     if (mysql && mysql_ping(mysql) == 0)
         /* The current server is up and running. */
@@ -113,6 +114,9 @@
         timeout = 5;
         mysql_options(mysql, MYSQL_OPT_CONNECT_TIMEOUT, (char*)&(timeout));
 
+	/* We want reconnect to happen. */
+	mysql_options(mysql, MYSQL_OPT_RECONNECT, &my_true);
+
         if (mysql_real_connect(mysql, mysql_servers->toks[n],
                 config_get_string("auth-mysql-username"),
                 password,
